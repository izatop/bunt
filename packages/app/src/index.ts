export * from "./Payload/index.js";
export * from "./Route/index.js";
export * from "./Request/index.js";
export * from "./TransformRequest/index.js";
export * from "./Application.js";
export * from "./interfaces.js";
