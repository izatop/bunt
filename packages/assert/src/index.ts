export * from "./Exception/index.js";
export * from "./assert.js";
