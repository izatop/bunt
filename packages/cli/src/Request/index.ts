export * from "./Headers.js";
export * from "./RequestCommand.js";
