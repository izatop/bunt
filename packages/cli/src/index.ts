export * from "./Transport/Headers";
export * from "./RequestCommand";
