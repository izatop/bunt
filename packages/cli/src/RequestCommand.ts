import {RequestAbstract} from "@typesafeunit/app";
import {assert, Console} from "@typesafeunit/util";
import {ReadStream} from "fs";
import {Headers} from "./Transport/Headers";

export class RequestCommand extends RequestAbstract {
    public readonly headers = new Headers([]);
    public readonly route: string;

    constructor() {
        super();
        const console = new Console();
        const [command] = console.argv.getArgs();
        assert(command, `Command should be defined`);
        this.route = command;
    }

    public createReadableStream(): ReadStream {
        return new ReadStream();
    }

    protected async write(response: string): Promise<void> {
        process.stdout.write(response + "\n");
    }
}
