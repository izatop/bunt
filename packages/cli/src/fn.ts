import {EqualMatcher, Route, RouteRuleArg} from "@bunt/app";
import {ActionAny, ActionFactory} from "@bunt/unit";

export function command<A extends ActionAny>(action: ActionFactory<A>, rule: RouteRuleArg<A>): Route<A> {
    return new Route<A>(EqualMatcher.factory, action, rule);
}
