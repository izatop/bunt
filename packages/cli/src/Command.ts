import {Application, IRoute, MatchRoute} from "@bunt/app";
import {ApplyContext, Context, ContextArg, IContext, unit} from "@bunt/unit";
import {RequestCommand} from "./Request";

export class Command<C extends IContext> extends Application<C> {
    public static async factory<C extends Context>(
        context: ContextArg<C>,
        routes: MatchRoute<C, IRoute>[] = []): Promise<Command<ApplyContext<C>>> {
        return new this(await unit<C>(context), routes);
    }

    public async handle(argv?: string[]): Promise<void> {
        const request = new RequestCommand(argv);
        await this.run(request);
    }
}
