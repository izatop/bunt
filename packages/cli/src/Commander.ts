import {Application, IRoute} from "@bunt/app";
import {Action, ApplyContext, ContextArg, Disposable, Heartbeat, IDisposable, IRunnable} from "@bunt/unit";
import {CommandContext, ICommandContext} from "./Context/CommandContext";
import {RequestCommand} from "./Request";

export class Commander<C extends ICommandContext> implements IRunnable, IDisposable {
    readonly #application: Application<C>;

    protected constructor(application: Application<C>) {
        this.#application = application;
    }

    public static async execute<C extends CommandContext>(
        context: ContextArg<C>, routes: IRoute<Action<C, any, IRunnable>>[] = []): Promise<Commander<ApplyContext<C>>> {
        const command = new this(await Application.factory(context, routes));
        return command.handle();
    }

    public async handle(): Promise<Commander<any>> {
        const {context} = this.#application;
        const request = new RequestCommand(context.program.args);

        return this.#application.run(request);
    }

    public async dispose(): Promise<void> {
        return;
    }

    public getHeartbeat(): Heartbeat {
        return Heartbeat.create(this, (fn) => Disposable.attach(this, fn));
    }
}
