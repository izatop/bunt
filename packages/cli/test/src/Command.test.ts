import {Resolver, RouteRule} from "@bunt/app";
import {Fields, Text} from "@bunt/input";
import {Commander} from "../../src";
import {BaseTestCommand} from "./action/BaseTestCommand";
import {command} from "./command";
import {BaseContext} from "./context/BaseContext";

test("Command", async () => {
    const baseTestCommand = command(BaseTestCommand, new RouteRule(
        "test",
        new Fields({name: Text}),
        new Resolver({name: ({context}) => context.args.getOption("name")}),
    ));

    const context = new BaseContext(["test", "--name=test"]);
    const action = await Commander.execute(context, [baseTestCommand]);
    expect(action).toMatchSnapshot();
});
