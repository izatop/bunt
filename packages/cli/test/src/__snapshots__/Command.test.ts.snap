// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command 1`] = `
BaseTestCommand {
  "context": BaseContext {
    "program": Program {
      "args": Argv {
        "args": Set {
          "test",
        },
        "flags": Map {},
        "options": Map {
          "name" => "test",
        },
      },
    },
  },
  "state": Object {
    "name": "test",
  },
}
`;
